hail/src/main/scala/is/hail/annotations/RegionPool.scala

```diff
@@ -1,10 +1,17 @@
 package is.hail.annotations
 
 import is.hail.utils._
+import org.apache.spark.TaskContext
 
 object RegionPool {
   private lazy val thePool: ThreadLocal[RegionPool] = new ThreadLocal[RegionPool]() {
-    override def initialValue(): RegionPool = RegionPool()
+    override def initialValue(): RegionPool = {
+      val pool = RegionPool()
+      TaskContext.get().addTaskCompletionListener { (_: TaskContext) =>
+        pool.close()
```

Comment:
I don't think this is valid. The executor thread stays alive to execute multiple tasks.

Instead of `.close()`, it should probably be `clear()`, I think, which reclaims all regions (does not free blocks, or only frees above some threshold) and frees all chunks.

Reply:
There's also a region pool on the driver JVM.

Reply:
Ah right. Do we just have to rely on java finalizers on the driver? ("Rely on" as a safety net. During a successful run, we should be manually freeing all allocations.)

Reply:
ExecuteContext.scoped can have a RegionPool.clear() on it too

Reply:
> we should be manually freeing all allocations.

We don't free blocks back to the operating system, because we assume that the next unit if work will have similar memory requirements.